cvxpy/tests/solver_test_helpers.py

```diff
@@ -110,27 +110,120 @@ def check_complementarity(self, places) -> None:
         #   attribute constraint.
         for con in self.constraints:
-            if isinstance(con, cp.constraints.PSD):
-                dv = con.dual_value
-                pv = con.args[0].value
-                comp = cp.scalar_product(pv, dv).value
+            if isinstance(con, (cp.constraints.Inequality,
+                                cp.constraints.Equality)):
+                comp = cp.scalar_product(con.expr, con.dual_value).value
             elif isinstance(con, (cp.constraints.ExpCone,
                                   cp.constraints.SOC,
-                                  cp.constraints.NonPos,
-                                  cp.constraints.Zero)):
+                                  cp.constraints.NonNeg,
+                                  cp.constraints.Zero,
+                                  cp.constraints.PSD,
+                                  cp.constraints.PowCone3D)):
                 comp = cp.scalar_product(con.args, con.dual_value).value
-            elif isinstance(con, cp.constraints.PowCone3D):
-                comp = cp.scalar_product(con.args[:3], con.dual_value).value
-            elif isinstance(con, (cp.constraints.Inequality,
-                                  cp.constraints.Equality)):
-                comp = cp.scalar_product(con.expr, con.dual_value).value
             elif isinstance(con, cp.constraints.PowConeND):
                 msg = '\nPowConeND dual variables not implemented;' \
                        + '\nSkipping complementarity check.'
                 warnings.warn(msg)
             else:
                 raise ValueError('Unknown constraint type %s.' % type(con))
             self.tester.assertAlmostEqual(comp, 0, places)
+            
+    def check_stationary_lagrangian(self, places) -> None:
+        L = self.prob.objective.expr
+        for con in self.constraints:
+            if isinstance(con, (cp.constraints.Inequality,
+                                cp.constraints.Equality)):
+                dual_var_value = con.dual_value
+                prim_var_expr = con.expr
+                if self.prob.objective.NAME == 'minimize':
+                    L = L + cp.scalar_product(dual_var_value, prim_var_expr)
+                else:
+                    L = L - cp.scalar_product(dual_var_value, prim_var_expr)
+            elif isinstance(con, (cp.constraints.ExpCone,
+                                  cp.constraints.SOC,
+                                  cp.constraints.Zero,
+                                  cp.constraints.NonNeg,
+                                  cp.constraints.PSD,
+                                  cp.constraints.PowCone3D)):
+                if self.prob.objective.NAME == 'minimize':
+                    L = L - cp.scalar_product(con.args, con.dual_value)
+                else:
+                    L = L + cp.scalar_product(con.args, con.dual_value)
+            else:
+                raise NotImplementedError()
```

Comment:
Instead of checking if the objective is ``minimize`` when processing constraints, just flip the sign on the objective function. I.e., do
```
    objective = self.prob.objective
    if objective.NAME == 'minimize':
        L = objective.expr
    else:
        L = -objective.expr
```